OCP 4.15: Setting SMT levels for Power in OpenShift Installer (#71019)
openshift/openshift-docs: modules/installation-ibm-power-vs-config-yaml.adoc

Reviewer: @alishaIBM  Check this line - https://github.com/openshift/openshift-docs/pull/71019/files#diff-4e884ac8d8b7162f130b92a57dc1a587d1acd51de51a2aefd1aa3dfb5369ce2fR348
That statement may not hold good for 4.15.

Refer to https://issues.redhat.com/browse/OCPSTRAT-959
Work is being done to remove "openshift-sdn" as an install-time option for newly-installed clusters for 4.15+. 

cc: @manojnkumar

Author: I have updated the PR. Thanks @juliemathew

- architecture: ppc64le
  hyperthreading: Enabled <3>
  name: worker
  platform:
    powervs:
      smtLevel: 4 <4>
  replicas: 3
controlPlane: <1> <2>
  architecture: ppc64le
  hyperthreading: Enabled <3>
  name: master
  platform:
    powervs:

>       smtLevel: 4 <4>

Reviewer 2: why this value is set to 4 here? not sure what's the basic convention in the OCP, is it default value or something else?

Author: Updated to the default value 8.
cc: @mkumatag @juliemathew

//====
//<6> You can optionally provide the `sshKey` value that you use to access the machines in your cluster.
//endif::openshift-origin[]

> <4> The SMTLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are from 1 to 8.

Reviewer 3: You cannot set it to any value between 1 and 8.  For e.g. 3, 5, 6, 7 etc.  Only Powers of 2 are allowed, 1, 2, 4, 8.

> <4> The smtLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are 1, 2, 3, 4, 5, 6, 7, 8, 'off' and 'on'. The smtLevel 'off' sets SMT to off and smtlevel 'on' sets SMT to 8 on the cluster nodes. For smtLevel 1 and 2, more number of processor cores need to be assigned to the cluster nodes while deploying. Ensure that the number of 'processors' set in the `install-config.yaml` file meets the requirements for deploying {product-title} successfully. One vCPU is equivalent to one physical core when simultaneous multithreading (SMT), or hyperthreading, is not enabled. When enabled, use the following formula to calculate the corresponding ratio: (threads per core × cores) × sockets = vCPUs.

Reviewer 3: We want to be more generic than just calling out smtLevel 1 and 2.  We should make a generic comment that when you use a smaller smtLevel, you may need to assign more cores.  Total vCPUs is computed as threads per core*number of cores*number of sockets.

Reviewer 2: This formula `(threads per core × cores) × sockets = vCPUs.` looks too generic, IMO we should use the powervm terminologies(need to dig deeper where to get that information)

Reviewer 2: also need to include what's the default value, add some examples what happens user doesn't specific anything and for the specific SMT value.

Author: > This formula (threads per core × cores) × sockets = vCPUs. looks too generic, IMO we should use the powervm terminologies(need to dig deeper where to get that information)

@mkumatag 
Referring to the output of `lscpu` command on the cluster nodes for terminology to be used in the documentation ([from QE task](https://github.ibm.com/redstack-power/project-mgmt/issues/3764#issuecomment-72125540)):
```
sh-5.1# lscpu
Architecture:            ppc64le
  Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-7
Model name:              POWER9 (architected), altivec supported
  Model:                 1.3 (pvr 004e 2103)
  Thread(s) per core:    8
  Core(s) per socket:    1
  Socket(s):             1
  Physical sockets:      16
  Physical chips:        1
  Physical cores/chip:   10
Virtualization features:
  Hypervisor vendor:     pHyp
  Virtualization type:   para
Caches (sum of all):
  L1d:                   64 KiB (2 instances)
  L1i:                   64 KiB (2 instances)
  L2:                    512 KiB (1 instance)
  L3:                    10 MiB (1 instance)
NUMA:
  NUMA node(s):          1
  NUMA node2 CPU(s):     0-7
```

Reviewer 2: interesting!

//====
//<6> You can optionally provide the `sshKey` value that you use to access the machines in your cluster.
//endif::openshift-origin[]

> <4> The smtLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are 1, 2, 4, 8, 'off' and 'on'. The default value is 8. The smtLevel 'off' sets SMT to off and smtlevel 'on' sets SMT to the default value 8 on the cluster nodes.

Reviewer 3: There is an inconsistency.  The statement in the bare-metal document states that the level can be any value between 1 and 8, and here we are specific to powers of 2.  We should be consistent.

Author: @manojnkumar My apologies. Updated the bare metal file with the correct value.

> <4> The smtLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are 1, 2, 4, 8, 'off' and 'on'. The default value is 8. The smtLevel 'off' sets SMT to off and smtlevel 'on' sets SMT to the default value 8 on the cluster nodes.

Reviewer 4: I also don't think the `off` and `on` are in backtics here either when they are meant to be.

Author: @kelbrown20 the single quotes for 'off' and 'on' are valid values. To include these values with backticks, can I write them as : `'off'` and `'on'` ? single quotes wrapped in backticks? Please confirm. Thanks.

Reviewer 4: Ohh so in the yaml it would be written like 'off' 'on', is that correct?

Author: > so in the yaml it would be written like 'off' 'on', is that correct?
Yes, correct. It should be written that way. FYI : @kelbrown20

Reviewer 4: Gotcha, my apologies for the confusion. I think it would still look good in the back tics like `'off'` `'on'` cuz whenever we have word values they're usually in backtics.

Author: @kelbrown20 I have updated the PR to have backticks for above values. Request you to check. Thanks.

Reviewer 4: Updates LGTM! sorry for the confusion here, thank you!

+
[NOTE]
====

> When simultaneous multithreading (SMT), or hyperthreading is not enabled, one vCPU is equivalent to one physical core. When enabled, total vCPUs is computed as (Thread(s) per core * Core(s) per socket) * Socket(s). The smtLevel controls the threads per core. Lower SMT levels may require additional cores to be assigned while deploying the cluster nodes. You can do this by setting the 'processors' in the `install-config.yaml` file to an appropriate value to meet the requirements for deploying {product-title} successfully.

Reviewer 4: ```suggestion
When simultaneous multithreading (SMT), or hyperthreading is not enabled, one vCPU is equivalent to one physical core. When enabled, total vCPUs is computed as: (Thread(s) per core * Core(s) per socket) * Socket(s). The smtLevel controls the threads per core. Lower SMT levels may require additional assigned cores when deploying the cluster nodes. You can do this by setting the 'processors' parameter in the `install-config.yaml` file to an appropriate value to meet the requirements for deploying {product-title} successfully.
```

====
<5> The cluster network plugin to install. The supported value is `OVNKubernetes`.
<6> The name of an existing resource group.
<7> Required. The installation program prompts you for this value.
ifdef::openshift-origin[]

> <8> You can optionally provide the `sshKey` value that you use to access the machines in your cluster.

Reviewer 4: Maybe
```suggestion
<8> Optional. You can provide the `sshKey` value that you use to access the machines in your cluster.
```

    cloudConnectionName: powervs-region-example-cloud-con-priv
    vpcSubnets:
    - powervs-region-example-subnet-1
    vpcRegion : vpc-region
    zone: powervs-zone
    serviceInstanceID: "powervs-region-service-instance-id"
publish: Internal <8>
pullSecret: '{"auths": ...}' <9>
sshKey: ssh-ed25519 AAAA... <10>
----
<1> If you do not provide these parameters and values, the installation program provides the default value.
<2> The `controlPlane` section is a single mapping, but the compute section is a sequence of mappings. To meet the requirements of the different data structures, the first line of the `compute` section must begin with a hyphen, `-`, and the first line of the `controlPlane` section must not. Both sections currently define a single machine pool. Only one control plane pool is used.
<3> Whether to enable or disable simultaneous multithreading, or `hyperthreading`. By default, simultaneous multithreading is enabled to increase the performance of your machines' cores. You can disable it by setting the parameter value to `Disabled`. If you disable simultaneous multithreading in some cluster machines, you must disable it in all cluster machines.

> <4> The SMTLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are from 1 to 8.

Reviewer 3: You cannot set it to any value between 1 and 8.  For e.g. 3, 5, 6, 7 etc.  Only Powers of 2 are allowed, 1, 2, 4, 8.

> <4> The smtLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are 1, 2, 3, 4, 5, 6, 7, 8, 'off' and 'on'. The smtLevel 'off' sets SMT to off and smtlevel 'on' sets SMT to 8 on the cluster nodes. For smtLevel 1 and 2, more number of processor cores need to be assigned to the cluster nodes while deploying. Ensure that the number of 'processors' set in the `install-config.yaml` file meets the requirements for deploying {product-title} successfully. One vCPU is equivalent to one physical core when simultaneous multithreading (SMT), or hyperthreading, is not enabled. When enabled, use the following formula to calculate the corresponding ratio: (threads per core × cores) × sockets = vCPUs.

Reviewer 3: Same comment as above, we cannot call out smtLevel 1 and 2.

    region: powervs-region
    vpcName: name-of-existing-vpc <7>
    vpcSubnets:
    - powervs-region-example-subnet-1
    vpcRegion : vpc-region
    zone: powervs-zone
    serviceInstanceGUID: "powervs-region-service-instance-guid"
publish: Internal <8>
pullSecret: '{"auths": ...}' <9>
sshKey: ssh-ed25519 AAAA... <10>
----
<1> If you do not provide these parameters and values, the installation program provides the default value.
<2> The `controlPlane` section is a single mapping, but the compute section is a sequence of mappings. To meet the requirements of the different data structures, the first line of the `compute` section must begin with a hyphen, `-`, and the first line of the `controlPlane` section must not. Both sections currently define a single machine pool. Only one control plane pool is used.
<3> Whether to enable or disable simultaneous multithreading, or `hyperthreading`. By default, simultaneous multithreading is enabled to increase the performance of your machines' cores. You can disable it by setting the parameter value to `Disabled`. If you disable simultaneous multithreading in some cluster machines, you must disable it in all cluster machines.

> <4> The smtLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are 1, 2, 4, 8, 'off' and 'on'. The default value is 8. The smtLevel 'off' sets SMT to off and smtlevel 'on' sets SMT to the default value 8 on the cluster nodes.

Reviewer 4: As well as where with the backtics

    vpcRegion : vpc-region
    vpcName: name-of-existing-vpc <7>
    vpcSubnets: <8>
    - powervs-region-example-subnet-1
    zone: powervs-zone
    serviceInstanceID: "powervs-region-service-instance-id"
credentialsMode: Manual
publish: External <9>
pullSecret: '{"auths": ...}' <10>
fips: false
sshKey: ssh-ed25519 AAAA... <11>
----
<1> If you do not provide these parameters and values, the installation program provides the default value.
<2> The `controlPlane` section is a single mapping, but the compute section is a sequence of mappings. To meet the requirements of the different data structures, the first line of the `compute` section must begin with a hyphen, `-`, and the first line of the `controlPlane` section must not. Both sections currently define a single machine pool. Only one control plane pool is used.
<3> Whether to enable or disable simultaneous multithreading, or `hyperthreading`. By default, simultaneous multithreading is enabled to increase the performance of your machines' cores. You can disable it by setting the parameter value to `Disabled`. If you disable simultaneous multithreading in some cluster machines, you must disable it in all cluster machines.

> <4> The SMTLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are from 1 to 8.

Reviewer 3: You cannot set it to any value between 1 and 8.  For e.g. 3, 5, 6, 7 etc.  Only Powers of 2 are allowed, 1, 2, 4, 8.

====
If you disable simultaneous multithreading, ensure that your capacity planning accounts for the dramatically decreased machine performance. Use larger machine types, such as `n1-standard-8`, for your machines if you disable simultaneous multithreading.
====

> <5> The SMTLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are from 1 to 8.

Reviewer 3: You cannot set it to any value between 1 and 8.  For e.g. 3, 5, 6, 7 etc.  Only Powers of 2 are allowed, 1, 2, 4, 8.

Reviewer: > You cannot set it to any value between 1 and 8. For e.g. 3, 5, 6, 7 etc. Only Powers of 2 are allowed, 1, 2, 4, 8.

As per @mjturek  all values from 1 to 8 are supported on Power.
QE tested and verified this with smtLevel set to 5 and 7. Cluster is stable in both cases.

With smtLevel set to 7, lscpu on the node shows
```
Architecture:            ppc64le
  Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-6
  Off-line CPU(s) list:  7
Model name:              POWER9 (architected), altivec supported
  Model:                 1.3 (pvr 004e 2103)
  Thread(s) per core:    7
  Core(s) per socket:    1
  Socket(s):             1
  Physical sockets:      16
  Physical chips:        1
  Physical cores/chip:   10
```

With smtLevel set to 5, lscpu on the nodes show this,
```
Architecture:            ppc64le
Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-4
  Off-line CPU(s) list:  5-7
Model name:              POWER9 (architected), altivec supported
  Model:                 1.3 (pvr 004e 2103)
  Thread(s) per core:    5
  Core(s) per socket:    1
  Socket(s):             1
  Physical sockets:      16
  Physical chips:        1
  Physical cores/chip:   10
```

fyi: @manojnkumar @alishaIBM

> <5> The smtLevel specifies the level of SMT to set to the control plane and compute machines. The supported values are 1, 2, 3, 4, 5, 6, 7, 8, 'off' and 'on'. The smtLevel 'off' sets SMT to off and smtlevel 'on' sets SMT to 8 on the cluster nodes. For smtLevel 1 and 2, more number of processor cores need to be assigned to the cluster nodes while deploying. Ensure that the number of 'processors' set in the `install-config.yaml` file meets the requirements for deploying {product-title} successfully. One vCPU is equivalent to one physical core when simultaneous multithreading (SMT), or hyperthreading, is not enabled. When enabled, use the following formula to calculate the corresponding ratio: (threads per core × cores) × sockets = vCPUs.

Reviewer 3: Same comment as above for smtLevel 1 and 2.
